app/referee/tasks.py

```diff
@@ -57,18 +57,62 @@ def run_and_score_submission(client, submission):
 def submit_submission_run(client, submission_run):
     delayed_conditional = dask.delayed(True)
     delayed_conditional = _run(submission_run, delayed_conditional)
+    if settings.VISUALIZE_DASK_GRAPH:
+        delayed_conditional.visualize(filename="task_graph.svg")
     future = client.submit(delayed_conditional.compute)  # pylint:disable=no-member
     logger.info("Future key: %s", future.key)
 
     dd.fire_and_forget(future)
     return future
 
 
+@dask.delayed(pure=False)
+def cache_containers(submission_run, delayed_conditional):
+
+    if not delayed_conditional:
+        return 1
+
+    container = submission_run.submission.container
+    aws_login_func = (
+        utils.get_aws_credential_function(container.uri)
+        if settings.LOGIN_TO_AWS
+        else None
+    )
+    submission_run.append(stdout=f"{container.uri}\n")
+    submission_run.append(stdout=f"{container.container_type}\n")
+
+    if settings.CONTAINER_FILES_ROOT:
+        container_save_path = (
+            settings.CONTAINER_FILES_ROOT
+            / f"{container.label.replace('/', '_')}_{container.tag}.sif"
```

Comment:
This seems risky to me. Since users can upload their own containers, this seems like a possible attack vector. We should construct the filename extremely carefully. Perhaps just use the container ID, user ID, and timestamp?

Reply:
fixed